firedrake/ml/pytorch/fem_operator.py

```diff
@@ -83,7 +83,7 @@ def backward(ctx, grad_output):
             adj_input = float(adj_input)
 
         # Compute adjoint model of `F`: delegated to pyadjoint.ReducedFunctional
-        adj_output = F.derivative(adj_input=adj_input)
+        adj_output = F.derivative(adj_input=adj_input, options={"riesz_representation": "L2"})
```

Comment:
@nbouziani With this PR, the functional derivatives will, by default, result from the L2 Riesz representation to map from the dual to the primal space. However, I am facing test failures in `firedrake/tests/regression/test_taylor_fem_operator.py`, specifically at `torch.autograd.gradcheck(G, x_P, eps=1e-6)`.
I have a couple of questions:

1. Should the `adj_output` be in the primal or dual space?

1. Do you have any idea why using `F.derivative(adj_input=adj_input, options={"riesz_representation": "L2"})` might lead to these test failures at `torch.autograd.gradcheck(G, x_P, eps=1e-6)`?

Reply:
1. It depends on the operation represented by `F`, but I don't think it matters in this case (see 2.)
2. I strongly suspect that's because PyTorch assumes `l2` product for the Riesz representation. I don't think we should default to `L2` here. I think we should either always impose `l2` or give that option to the `fem_operator` and default to `l2`. It's not clear to me why considering other inner products would make sense, given that `l2` is assumed throughout the AD calculations in PyTorch.